Teach `ActiveJob` to set configs on itself

sambostock opened the pull request:
### Summary

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

Previously configs of the form `config.active_job.X` were only forwarded to `ActiveJob::Base`. This teaches Active Job to set them on `ActiveJob` directly instead, if the setter exists.


### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

<!--
Note: Please avoid making *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.
Create a pull request when it is ready for review and feedback
from the Rails team :).
-->

This is a follow up to #45618, which introduces the first config set on `ActiveJob` directly. @adrianna-chang-shopify pointed out that despite having set the config

```ruby
# config/application.rb
config.active_job.use_big_decimal_serializer = true
```

the config was still set to the default value after the application was booted:

```ruby
Rails.application.config.active_job.use_big_decimal_serializer # => true
ActiveJob.use_big_decimal_serializer                           # => false
```

In #45618, I had considered if it was more appropriate to set the config on `ActiveJob::Base`. While that might have allowed individual job classes to opt or out of the serializer (which is of questionable value), it would have required a change to the `Arguments.serialize` API to inject the job class. I assumed that the config would be automatically forwarded, without realizing that each framework is responsible for setting their config independently (perhaps an opportunity for helper extraction).

sambostock commented on activejob/CHANGELOG.md:
```diff
@@ -1,3 +1,13 @@
+*   Teach initializer to set configs directly on `ActiveJob`
+
+    Previously, configs of the form `config.active_job.NAME` were automatically
+    set on `ActiveJob::Base` if it responds to `NAME=`.
+
+    To support the new `use_big_decimal_serializer` flag, setting configs on
+    `ActiveJob` is supported as well.
+
+    *Adrianna Chang* and *Sam Bostock*
+
```
I am unsure of if this needs a CHANGELOG entry, as it doesn't _really_ impact consumers, given it's only changing to support an unreleased feature.

That said, it is possible that someone tracking `main` thinks they have adopted `use_big_decimal_serializer`, when in fact they haven't because the config wasn't being set properly.

casperisfine replied:
Yeah I don't think it's needed.

sambostock replied:
Okay, removed 👍

sambostock commented on activejob/lib/active_job/railtie.rb:
```diff
@@ -34,7 +34,11 @@ class Railtie < Rails::Railtie # :nodoc:
         options.each do  |k, v|
           k = "#{k}="
-          send(k, v) if respond_to? k
+          if ActiveJob.respond_to?(k)
+            ActiveJob.send(k, v)
+          elsif respond_to? k
+            send(k, v)
+          end
```
This does change the behaviour if the setter exists on **both `ActiveJob` and `ActiveJob::Base`, as it will now no longer be set on `Base`. We could alternatively do something like
```suggestion
          ActiveJob.send(k, v) if ActiveJob.respond_to? k
          send(k, v) if respond_to? k
```

sambostock replied:
I should also note I wasn't able to find any settings for the Railtie. Are there any? If so, it may be worth updating them.

casperisfine replied:
Nah it's fine. It works like this in others gems line Active Record.

skipkayhil commented:
Probably worth testing in `railties/test/application/configuration_test.rb` like https://github.com/rails/rails/blob/b01027fc2faf0f9c919152bdf0c4d5db56ea41fb/railties/test/application/configuration_test.rb#L2645

sambostock commented on railties/test/application/configuration_test.rb:
```diff
@@ -2654,6 +2654,28 @@ class ::DummySerializer < ActiveJob::Serializers::ObjectSerializer; end
+
+      app "test"
+
+      # When loaded, ActiveJob::Base triggers the :active_job load hooks, which is where config is attached.
+      # Referencing the constant auto-loads it.
+      ActiveJob::Base
```
This took a **long** time to figure out, as the assertion would just fail without it. It was particularly hard to test because the tests run in parallel with their output interleaved, and this suite would crash due to running out of file descriptors... (See #45667 😅)

It feels kind of weird that we have to do this. Nothing in the implementation `BigDecimalSerializer`, its use in `Assertions`, or let alone the configuration code references `ActiveJob::Base`, so it's weird that we have to reference it to get this to work. Is there a better way to do this, or is this acceptable?

I wonder if maybe I should have just set the config on `ActiveJob::Base` instead? That just felt like it attached it to a job, which `Arguments` isn't. 🤔

casperisfine replied:
It probably shouldn't be set on `on_load`, Active Record does it on `after_initialize`, might as well be consistent: https://github.com/rails/rails/blob/ab325a920986187f9d1c13bdfba77682ab6f6503/activerecord/lib/active_record/railtie.rb#L226-L238

sambostock replied:
Okay, I've mirrored the approach, including the "double copying" for consistency. `ActiveRecord` does it for backwards compatibility though, which we don't technically need to support here (since `use_big_decimal_serializer` is the first config on `ActiveJob` directly).

sambostock reviewed (commented):
Ahh, thank you @skipkayhil. I've added tests for the default value, and overriding it.

rafaelfranca commented on activejob/lib/active_job/railtie.rb:
```diff
@@ -25,16 +25,29 @@ class Railtie < Rails::Railtie # :nodoc:
       options = app.config.active_job
       options.queue_adapter ||= :async
 
+      options.after_initialize do
```
Should not this be `app.config`? And why do it twice?

sambostock replied:
@byroot recommended [copying the approach from `ActiveRecord::Railtie`](https://github.com/rails/rails/blob/bfa3a5baabec97855d77615645bb6757f19a4502/activerecord/lib/active_record/railtie.rb#L223-L268) for consistency, which is why we're doing it twice. I'm glad to change this to whatever is determined to be the correct way to do it.

However, I do see that I misread, and `ActiveRecord::Railtie` is calling `after_initialize` on `config`, not `configs`, so I'll fix that. It is clearer here with the (existing) name `options`, which stands out as different.

byroot commented:
@sambostock can you also add a fix for the issue @rafaelfranca noticed in https://github.com/rails/rails/pull/45618#discussion_r931353133 ? It can go in the same commit as this.

sambostock committed:
Teach ActiveJob to set configs on itself

Previously configs of the form `config.active_job.X` were only forwarded to
`ActiveJob::Base`. This teaches Active Job to set them on `ActiveJob` directly
instead, if the setter exists.

For consistency, this more or less mirrors the way that Active Record does it.

Co-authored-by: Adrianna Chang <adrianna.chang@shopify.com>
Co-authored-by: Sam Bostock <sam.bostock@shopify.com>

---

Fix use_big_decimal_serializer Rails 7.1 default

This config should be enabled for new Rails 7.1 apps, or apps that have updated
their config to `load_defaults 7.1`, not disabled.

This also clarifies the config accessor comment.

---

Add contributor documentation comment to load_defaults

The process for introducing a change in behavior in Rails can be confusing to
new contributors, so a comment is added roughly explaining how to do so, and
what belongs in `load_defaults` and `new_framework_defaults`.

This comment is aimed at contributors, not consumers, so it is added within the
method, rather than above it.

sambostock commented:
@byroot done! I've also added the contributor documentation comment, as discussed. Hopefully I got it all correct.

byroot commented:
Thanks Sam. I'll try to review this later today, in the meantime can you squash your commits? Otherwise I can't merge.

byroot commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -87,6 +87,18 @@ def initialize(*)
+        # 1. Add an accessor on the target object (e.g. the ActiveJob class for global Active Job config).
+        # 2. Set a default value there preserving existing behavior for existing applications.
+        # 3. In the section below corresponding to the next release of Rails, set the new value.
+        # 4. Add a commented out section in the new_framework_defaults file, setting the new value.
+        # 5. Update the guide on configuration.
```
```suggestion
        # 4. Add a commented out section in the new_framework_defaults template, setting the new value.
        # 5. Update the guide in `configuring.md`
```

sambostock replied:
Updated. I also realized I omitted the step of actually implementing the change, and its corresponding removal step, so I added that too.

Commits squashed too.
